pkg/controllers/cspc-controller/scale.go

```diff
@@ -0,0 +1,184 @@
+			runtime.HandleError(errors.Wrapf(err, "Pool provisioning failed for %d/%d for cstorpoolcluster %s", poolCount, pendingPoolCount, cspc.Name))
+		} else {
+			message := fmt.Sprintf("Pool Provisioned %d/%d ", poolCount, pendingPoolCount)
+			c.recorder.Event(cspc, corev1.EventTypeNormal, "Create", message)
+			klog.Infof("Pool provisioned successfully %d/%d for cstorpoolcluster %s", poolCount, pendingPoolCount, cspc.Name)
+		}
+	}
+	return nil
+}
+
+// CreateCSPI creates CSPI
+func (c *Controller) CreateCSPI(cspc *cstor.CStorPoolCluster) error {
+	ac, err := algorithm.NewBuilder().
+		WithCSPC(cspc).
+		WithNameSpace(cspc.Namespace).
+		WithKubeClient(c.kubeclientset).
+		WithOpenEBSClient(c.clientset).
+		Build()
+	if err != nil {
+		return err
+	}
+	cspi, err := ac.GetCSPSpec()
+	if err != nil {
+		return err
+	}
+	// The cpsi variable is written back here, This is important as cspi uid is passed to pool deployment
+	// The uid does not exist before cspi creation.
+	cspi, err = c.GetStoredCStorVersionClient().CStorPoolInstances(cspc.Namespace).Create(cspi)
+
+	if err != nil {
+		return err
+	}
+	err=c.CreateCSPIDeployment(cspc,cspi)
+	if err!=nil{
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) createDeployForCSPList(cspc *cstor.CStorPoolCluster, cspList []cstor.CStorPoolInstance) {
+	for _, cspObj := range cspList {
+		cspObj := cspObj
+		err := c.CreateCSPIDeployment(cspc,&cspObj)
+		if err != nil {
+			message := fmt.Sprintf("Failed to create pool deployment for CSP %s: %s", cspObj.Name, err.Error())
+			c.recorder.Event(cspc, corev1.EventTypeWarning, "PoolDeploymentCreate", message)
+			runtime.HandleError(errors.Errorf("Failed to create pool deployment for CSP %s: %s", cspObj.Name, err.Error()))
+		}
+	}
+}
+// CreateStoragePool creates the required resource to provision a cStor pool
+func (c *Controller) CreateCSPIDeployment(cspc *cstor.CStorPoolCluster,cspi *cstor.CStorPoolInstance) error {
+	ac, err := algorithm.NewBuilder().
+		WithCSPC(cspc).
+		WithNameSpace(cspc.Namespace).
+		WithKubeClient(c.kubeclientset).
+		WithOpenEBSClient(c.clientset).
+		Build()
+	if err != nil {
+		return err
+	}
+	deploy,err:=ac.GetPoolDeploySpec(cspi)
+	if err != nil {
+		return err
+	}
+	_,err=c.kubeclientset.AppsV1().Deployments(cspi.Namespace).Create(deploy)
+	if err!=nil{
+		return err
+	}
+	return nil
+}
+
+// DownScalePool deletes the required pool.
+func (c *Controller) ScaleDown(cspc *cstor.CStorPoolCluster) error {
+	orphanedCSP, err := c.getOrphanedCStorPools(cspc)
+	if err != nil {
+		c.recorder.Event(cspc, corev1.EventTypeWarning,
+			"DownScale", "Pool downscale failed "+err.Error())
+		return errors.Wrap(err, "could not get orphaned CSP(s)")
+	}
+	for _, cspiName := range orphanedCSP {
+		c.recorder.Event(cspc, corev1.EventTypeNormal,
+			"DownScale", "De-provisioning pool "+cspiName)
+
+		// TODO : As part of deleting a CSP, do we need to delete associated BDCs ?
```

Comment:
We can remove TODO because of `cleanupCSPIResources` is doing a cleanup work of CSPI.